src/python/pants/goal/run_tracker.py

```diff
@@ -414,21 +414,69 @@ def shutdown_worker_pool(self):
     """
     SubprocPool.shutdown(self._aborted)
 
-  def report_target_info(self, scope, target, key, val):
+  @staticmethod
+  def create_dict_with_nested_keys_and_val(all_keys, value, index):
+    """ Recursively constructs a nested dictionary with the keys pointing to the value.
+
+    :param list of string all_keys: A list of keys to be nested as a dictionary.
+    :param dict or string value: The value of the information being stored.
+    :param int index: The index of the last element of the list of keys.
+    :return: dict of nested keys leading to the value.
+    """
+    if index > 0:
+      new_val = {all_keys[index]: value}
+      return RunTracker.create_dict_with_nested_keys_and_val(all_keys, new_val, index - 1)
+    elif index == 0:
+      return {all_keys[index]: value}
+
+
+  @staticmethod
+  def merge_list_of_keys_into_dict(data, keys, value, index):
+    """ Recursively merge list of keys that points to the given value into data.
+
+    :param dict data: Dictionary to be updated.
+    :param list of string keys: The keys that point to where the value should be stored.
+           Will recursively find the correct place to store in the nested dicts.
+    :param primitive value: The value of the information being stored.
+    :param int index: The index into the list of keys (starting form the beginning). This should
+           always be 0 when being called outside of this function.
+    """
+    if len(keys) == 0:
+      raise ValueError('Keys must contain at least one key')
+    if len(keys) < 2 or not data:
+      new_data_to_add = RunTracker.create_dict_with_nested_keys_and_val(keys, value, len(keys) - 1)
+      data.update(new_data_to_add)
+      return
+    else:
+      this_keys_contents = data.get(keys[index])
+      if this_keys_contents:
+        if isinstance(this_keys_contents, dict):
+          new_keys = keys[index+1:]
+          RunTracker.merge_list_of_keys_into_dict(this_keys_contents, new_keys, value, 0)
+        else:
+          new_keys = keys[index+1:]
+          new_data_to_add = RunTracker.create_dict_with_nested_keys_and_val(new_keys, value, len(new_keys) - 1)
```

Comment:
I was slicing the the arrays because it simplifies the `create_dict_with_nested_keys_and_val` method. Since it that method starts from the end of the list and works forward, if we slice the array, we don't need to worry about passing in another integer that is the highest key we want in the nested dict. 

But you are totally right that the recursive call to `merge_list_of_keys_into_dict` should just increase the index and not slice the keys.